### PR TITLE
metadata bounds asset checks

### PR DESCRIPTION
**Pull request by johannkm**
Add an experimental `build_metadata_bounds_checks` factory for asserting that a numeric metadata value is within a range.

```python
@asset
def my_asset():
  return MaterializeResult(metadata={"my_metadata": 5})

checks = build_metadata_bounds_checks(
  assets=[my_asset], metadata_key="my_metadata", min_value=1, max_value=10
)

### CODE REVIEW

**Comment by johannkm** [comment]
* **#21630** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21630?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @johannkm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by sryza — DISMISSED** [review]
excited for this!

File: python_modules/dagster/dagster/_core/definitions/asset_check_factories/metadata_range_checks.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        specs=[
+            AssetCheckSpec(
+                f"{metadata_key}_range_check",
+                asset=asset_key,
+                description=description,
+            )
+            for asset_key in asset_keys
+        ],
+        can_subset=True,
+        name=unique_id_from_asset_keys(list(asset_keys)),
```
[the remark this answers is not in the document]
Comment (johannkm) [reply python_modules/dagster/dagster/_core/definitions/asset_check_factories/metadata_range_checks.py:117]: good point

File: python_modules/dagster/dagster/_core/definitions/asset_check_factories/metadata_range_checks.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+from ..asset_check_spec import AssetCheckKey, AssetCheckSeverity, AssetCheckSpec
+from ..asset_checks import AssetChecksDefinition
+from ..asset_key import AssetKey, CoercibleToAssetKey
+from ..assets import AssetsDefinition, SourceAsset
+from ..decorators.asset_check_decorator import multi_asset_check
+from .utils import unique_id_from_asset_keys
+
+
+@experimental
+def build_metadata_range_checks(
```
[the remark this answers is not in the document]
Comment (johannkm) [reply python_modules/dagster/dagster/_core/definitions/asset_check_factories/metadata_range_checks.py:19]: I like `build_metadata_limit_checks`

File: python_modules/dagster/dagster/_core/definitions/asset_check_factories/metadata_range_checks.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                f"Value `{value}` is less than {'or equal to ' if strict_min else ''}{min_value}",
+            )
+        if max_value is not None and (value >= max_value if strict_max else value > max_value):
+            return (
+                False,
+                f"Value `{value}` is greater than {'or equal to ' if strict_max else ''}{max_value}",
+            )
+        return True, f"Value `{value}` is within range"
+
+    @multi_asset_check(
```
[the remark this answers is not in the document]
Comment (johannkm) [reply python_modules/dagster/dagster/_core/definitions/asset_check_factories/metadata_range_checks.py:107]: check factory factory

[the remark this answers is not in the document]
Comment (johannkm) [reply python_modules/dagster/dagster/_core/definitions/asset_check_factories/metadata_range_checks.py]: moved schema checks and this to a shared factory, leaving freshness at the moment bc that has a more specific util for the 3 freshness check types

**Comment by sryza** [comment]
As I think about more places where these names will show up – e.g. a button in the UI next to a metadata value that allows creating a check from it –, I feel less convinced on "limit".  "Limit" feels like it has the connotation of "rate limit" or "select * from x limit 5".

This leads me to lean more towards something like "bound" or "threshold".  Maybe worth getting the opinions of the dev rel folks?

**Comment by johannkm** [comment]
updated to *bounds*

File: docs/sphinx/sections/api/apidocs/asset-checks.rst [hunk 4]
```diff
@@ -34,3 +34,5 @@ Dagster allows you to define and execute checks on your software-defined assets.
 .. autofunction:: build_sensor_for_freshness_checks 
 
 .. autofunction:: build_column_schema_change_checks
+
+.. autofunction:: build_metadata_range_checks
```
Comment (sryza) [inline docs/sphinx/sections/api/apidocs/asset-checks.rst:38]:
```suggestion
.. autofunction:: build_metadata_bounds_checks
```

File: python_modules/dagster-test/dagster_test/toys/asset_checks.py [hunk 5]
```diff
@@ -435,6 +437,23 @@ def downstream_asset():
 )
 
 
+@asset(
+    group_name="asset_checks",
```
Comment (sryza) [inline python_modules/dagster-test/dagster_test/toys/asset_checks.py:441]:
```suggestion
    group_name="asset_checks"
```

File: python_modules/dagster/dagster/_core/definitions/asset_check_factories/utils.py [hunk 6]
```diff
@@ -262,3 +272,27 @@ def inner(fn: MultiAssetCheckFunction) -> AssetChecksDefinition:
         )(fn)
 
     return inner
+
+
+def build_multi_asset_check_factory(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/asset_check_factories/utils.py:277]:
```suggestion
def build_multi_asset_check(
```

It's building a multi-asset check, not a factory, right?

**Review by sryza — DISMISSED** [review]
This looks very close

**Review by sryza — APPROVED** [review]
Nice

**Pull request merged**